File: Coding/python/design.py
"""Idiomatic Python 3 port of java/design.java (original package code.ds).

Object-oriented design exercise: a parking lot with small / compact / large
slots, several vehicle types, and park / un-park bookkeeping.

The Java ``Slot`` class had a field ``isOccupied`` and a method
``isOccupied()``; in Python the field is renamed ``_occupied`` and the accessor
stays ``is_occupied()``. Note the source's vehicles never override the
``canFit*`` predicates, so every ``canFit`` returns ``False`` and ``park``
always returns the sentinel token ``-1`` -- this behaviour is preserved. The
Java source has no ``main``; a demonstrative one is added.
"""


# ---------------------------------------------------------------------------
# Vehicles
# ---------------------------------------------------------------------------
class Vehicle:
    _counter = 0  # supplies a stable identity hash (Java Object.hashCode)

    def __init__(self):
        self.licence_plate = None
        Vehicle._counter += 1
        self._hash = Vehicle._counter

    def can_fit_in_small_slot(self):
        return False

    def can_fit_in_compact_slot(self):
        return False

    def can_fit_in_large_slot(self):
        return False

    def hash_code(self):
        return self._hash


class Car(Vehicle):
    pass


class Motorcycle(Vehicle):
    pass


class Bus(Vehicle):
    pass


class Truck(Vehicle):
    pass


# ---------------------------------------------------------------------------
# Slots
# ---------------------------------------------------------------------------
class Slot:
    def __init__(self, slot_number):
        self._occupied = False  # Java field isOccupied (renamed)
        self.slot_number = slot_number

    def is_occupied(self):
        return self._occupied

    def get_slot_number(self):
        return self.slot_number

    def park(self):
        self._occupied = True

    def un_park(self):
        self._occupied = False

    def __eq__(self, other):
        return isinstance(other, Slot) and other.slot_number == self.slot_number

    def __hash__(self):
        h = 5
        h = 53 * h + self.slot_number
        return h


class SmallSlot(Slot):
    pass


class CompactSlot(Slot):
    pass


class LargeSlot(Slot):
    pass


# ---------------------------------------------------------------------------
# Parking lot
# ---------------------------------------------------------------------------
class ParkingLot:
    NUMBER_OF_SMALL_SLOTS = 10
    NUMBER_OF_COMPACT_SLOTS = 10
    NUMBER_OF_LARGE_SLOTS = 10

    def __init__(self):
        self.small_slots = []   # stacks
        self.compact_slots = []
        self.large_slots = []
        self._create_slots()
        self.occupied_slots = {}  # token -> Slot

    def _create_slots(self):
        for i in range(1, self.NUMBER_OF_SMALL_SLOTS + 1):
            self.small_slots.append(SmallSlot(i))
        for i in range(1, self.NUMBER_OF_COMPACT_SLOTS + 1):
            self.compact_slots.append(CompactSlot(i))
        for i in range(1, self.NUMBER_OF_LARGE_SLOTS + 1):
            self.large_slots.append(LargeSlot(i))

    def park(self, vehicle):
        slot = None
        unique_token = -1

        if vehicle.can_fit_in_small_slot():
            slot = self.small_slots.pop()
            if slot is not None:
                unique_token = self._park_helper(slot, vehicle)
        if vehicle.can_fit_in_compact_slot():
            slot = self.compact_slots.pop()
            if slot is not None:
                unique_token = self._park_helper(slot, vehicle)
        if vehicle.can_fit_in_large_slot():
            slot = self.large_slots.pop()
            if slot is not None:
                unique_token = self._park_helper(slot, vehicle)
        return unique_token

    def un_park(self, unique_token):
        slot = self.occupied_slots.get(unique_token)
        slot.un_park()
        if isinstance(slot, SmallSlot):
            self.small_slots.append(slot)
        self.occupied_slots.pop(unique_token, None)

    def _park_helper(self, slot, vehicle):
        slot.park()
        unique_token = vehicle.hash_code() * 43
        self.occupied_slots[unique_token] = slot
        return unique_token


# ---------------------------------------------------------------------------
# Floors / areas (placeholder scaffolding from the source)
# ---------------------------------------------------------------------------
class ParkingFloor:
    def __init__(self):
        self.available_car_parking_slots = []
        self.available_bike_parking_slots = []
        self.occupied_car_parking_slots = []
        self.occupied_bike_parking_slots = []

    def get_number_of_available_car_parking_slots(self):
        return []

    def get_number_of_available_bike_parking_slots(self):
        return []


class ParkingArea:
    def __init__(self):
        self.parking_floors = []


if __name__ == "__main__":
    lot = ParkingLot()
    car = Car()
    token = lot.park(car)
    # -1: Car never overrides the canFit* predicates (faithful to the source).
    print("token:", token)

    slot = SmallSlot(7)
    slot.park()
    print("slot", slot.get_slot_number(), "occupied:", slot.is_occupied())

File: Coding/golang/design/design.go
package main

// Ported from Coding/java/design.java (original package code.ds).
// Parking-lot OO design. The Java file had no main; a small demo is added.

import "fmt"

const (
	numberOfSmallSlots   = 10
	numberOfCompactSlots = 10
	numberOfLargeSlots   = 10
)

func main() {
	lot := newParkingLot()
	car := newCar()
	token := lot.park(car)
	// canFit* all return false in the reference design, so park yields -1.
	fmt.Println("Parked vehicle, token:", token)
}

// Vehicle is the abstract base for things that can be parked.
type Vehicle interface {
	canFitInSmallSlot() bool
	canFitInCompactSlot() bool
	canFitInLargeSlot() bool
	hashCode() int64
}

type baseVehicle struct {
	licencePlate string
	id           int64
}

func (v *baseVehicle) canFitInSmallSlot() bool   { return false }
func (v *baseVehicle) canFitInCompactSlot() bool { return false }
func (v *baseVehicle) canFitInLargeSlot() bool   { return false }
func (v *baseVehicle) hashCode() int64           { return v.id }

var vehicleCounter int64

func nextVehicleID() int64 {
	vehicleCounter++
	return vehicleCounter
}

type Car struct{ baseVehicle }
type Motorcycle struct{ baseVehicle }
type Bus struct{ baseVehicle }
type Truck struct{ baseVehicle }

func newCar() *Car               { return &Car{baseVehicle{id: nextVehicleID()}} }
func newMotorcycle() *Motorcycle { return &Motorcycle{baseVehicle{id: nextVehicleID()}} }
func newBus() *Bus               { return &Bus{baseVehicle{id: nextVehicleID()}} }
func newTruck() *Truck           { return &Truck{baseVehicle{id: nextVehicleID()}} }

// Slot is the abstract base for a parking slot.
type Slot interface {
	park()
	unPark()
	isOccupied() bool
	getSlotNumber() int
}

type baseSlot struct {
	occupied   bool // renamed from isOccupied to avoid clashing with isOccupied()
	slotNumber int
}

func (s *baseSlot) isOccupied() bool   { return s.occupied }
func (s *baseSlot) getSlotNumber() int { return s.slotNumber }
func (s *baseSlot) park()              { s.occupied = true }
func (s *baseSlot) unPark()            { s.occupied = false }

func (s *baseSlot) equals(o Slot) bool {
	return o.getSlotNumber() == s.slotNumber
}

func (s *baseSlot) hashCode() int {
	hash := 5
	hash = 53*hash + s.slotNumber
	return hash
}

type SmallSlot struct{ baseSlot }
type CompactSlot struct{ baseSlot }
type LargeSlot struct{ baseSlot }

func newSmallSlot(n int) *SmallSlot     { return &SmallSlot{baseSlot{slotNumber: n}} }
func newCompactSlot(n int) *CompactSlot { return &CompactSlot{baseSlot{slotNumber: n}} }
func newLargeSlot(n int) *LargeSlot     { return &LargeSlot{baseSlot{slotNumber: n}} }

// ParkingLot manages slots of each size as stacks.
type ParkingLot struct {
	occupiedSlots map[int64]Slot
	smallSlots    []Slot
	compactSlots  []Slot
	largeSlots    []Slot
}

func newParkingLot() *ParkingLot {
	p := &ParkingLot{occupiedSlots: map[int64]Slot{}}
	p.createSlots()
	return p
}

func (p *ParkingLot) createSlots() {
	for i := 1; i <= numberOfSmallSlots; i++ {
		p.smallSlots = append(p.smallSlots, newSmallSlot(i))
	}
	for i := 1; i <= numberOfCompactSlots; i++ {
		p.compactSlots = append(p.compactSlots, newCompactSlot(i))
	}
	for i := 1; i <= numberOfLargeSlots; i++ {
		p.largeSlots = append(p.largeSlots, newLargeSlot(i))
	}
}

// popSlot pops the top slot off a stack, or nil if empty.
func popSlot(stack *[]Slot) Slot {
	s := *stack
	if len(s) == 0 {
		return nil
	}
	top := s[len(s)-1]
	*stack = s[:len(s)-1]
	return top
}

func (p *ParkingLot) park(vehicle Vehicle) int64 {
	var slot Slot
	var uniqueToken int64 = -1

	if vehicle.canFitInSmallSlot() {
		slot = popSlot(&p.smallSlots)
		if slot != nil {
			uniqueToken = p.parkHelper(slot, vehicle)
		}
	}
	if vehicle.canFitInCompactSlot() {
		slot = popSlot(&p.compactSlots)
		if slot != nil {
			uniqueToken = p.parkHelper(slot, vehicle)
		}
	}
	if vehicle.canFitInLargeSlot() {
		slot = popSlot(&p.largeSlots)
		if slot != nil {
			uniqueToken = p.parkHelper(slot, vehicle)
		}
	}
	return uniqueToken
}

func (p *ParkingLot) unPark(uniqueToken int64) {
	slot := p.occupiedSlots[uniqueToken]
	if slot == nil {
		return
	}
	slot.unPark()
	if _, ok := slot.(*SmallSlot); ok {
		p.smallSlots = append(p.smallSlots, slot)
	}
	delete(p.occupiedSlots, uniqueToken)
}

func (p *ParkingLot) parkHelper(slot Slot, vehicle Vehicle) int64 {
	slot.park()
	uniqueToken := vehicle.hashCode() * 43
	p.occupiedSlots[uniqueToken] = slot
	return uniqueToken
}

// ParkingFloor groups available/occupied slots per floor.
type ParkingFloor struct {
	availableCarParkingSlots  []Slot
	availableBikeParkingSlots []Slot
	occupiedCarParkingSlots   []Slot
	occupiedBikeParkingSlots  []Slot
}

func (f *ParkingFloor) getNumberOfAvailableCarParkingSlots() []Slot {
	return []Slot{}
}

func (f *ParkingFloor) getNumberOfAvailableBikeParkingSlots() []Slot {
	return []Slot{}
}

// ParkingArea groups multiple floors.
type ParkingArea struct {
	parkingFloors []*ParkingFloor
}
